jinahub/rankers/DPRReaderRanker/dpr_reader.py

```diff
@@ -0,0 +1,231 @@
+        self.device = device
+        self.max_length = max_length
+        self.num_spans_per_match = num_spans_per_match
+
+        if not base_tokenizer_model:
+            base_tokenizer_model = pretrained_model_name_or_path
+
+        self.tokenizer = DPRReaderTokenizerFast.from_pretrained(base_tokenizer_model)
+        self.model = DPRReader.from_pretrained(pretrained_model_name_or_path)
+
+        self.model = self.model.to(torch.device(self.device)).eval()
+
+        self.default_traversal_paths = default_traversal_paths
+        self.default_batch_size = default_batch_size
+
+    @requests
+    def rank(
+        self, docs: Optional[DocumentArray], parameters: dict, **kwargs
+    ) -> DocumentArray:
+        """
+        Extracts answers from existing matches, (re)ranks them, and replaces the current
+        matches with extracted answers.
+
+        For each match ``num_spans_per_match`` of answers will be extracted, which
+        means that the new matches of the document will have a length of previous
+        number of matches times ``num_spans_per_match``.
+
+        The new matches will be have a score called ``relevance_score`` saved under
+        their scores. They will also have a tag ``span_score``, which refers to their
+        span score, which is used to rank answers that come from the same match.
+
+        If you specified ``title_tag_key`` at initialization, the tag ``title`` will
+        also be added to the new matches, and will equal the title of the match from
+        which they were extracted.
+
+        :param docs: Documents whose matches to re-rank (specifically, the matches of
+            the documents on the traversal paths will be re-ranked). The document's
+            ``text`` attribute is taken as the question, and the ``text`` attribute
+            of the matches as the context. If you specified ``title_tag_key`` at
+            initialization, the matches must also have a title (under this tag).
+        :param parameters: dictionary to define the ``traversal_path`` and the
+            ``batch_size``. For example::
+            parameters={'traversal_paths': ['r'], 'batch_size': 10}
+        """
+        if not docs:
+            return None
+
+        new_matches = []
+        old_matches = []
+
+        docs_list = []
+        old_match_ref_doc_dict = {}
+        new_match_ref_doc_dict = {}
+
+        # Perpare a list of docs to add matches to later and extract their
+        # matches, plus build a reference to their reference docs
+        trav_paths = parameters.get('traversal_paths', self.default_traversal_paths)
+        for ind, doc in enumerate(docs.traverse_flat(trav_paths)):
+            if not doc.text:
+                raise ValueError(
+                    f'No question (text) found for document with id {doc.id}.'
+                    ' DPRReaderRanker requires a question from the main document'
+                    ' and context (text) from its matches.'
+                )
```

Comment:
I suggest skipping the docs without `text`. Instead of raising an exception and stop working, we can raise a warning and continue processing the others.
```suggestion
            if not doc.text:
                raise ValueError(
                    f'No question (text) found for document with id {doc.id}.'
                    ' DPRReaderRanker requires a question from the main document'
                    ' and context (text) from its matches.'
                )
```